File: algorithmic-techniques/Divide and Conquer/C++/code.cpp
#include <iostream>
#include <vector>
#include <climits>

void merge(
    int* array,
    int left,
    int middle,
    int right
)
{
    int leftSize = middle - left + 1;
    int rightSize = right - middle;

    std::vector<int> leftArray(leftSize);
    std::vector<int> rightArray(rightSize);

    for (int i = 0; i < leftSize; i++)
    {
        leftArray[i] = array[left + i];
    }

    for (int i = 0; i < rightSize; i++)
    {
        rightArray[i] = array[middle + 1 + i];
    }

    int i = 0;
    int j = 0;
    int k = left;

    while (i < leftSize && j < rightSize)
    {
        if (leftArray[i] <= rightArray[j])
        {
            array[k] = leftArray[i];
            i++;
        }
        else
        {
            array[k] = rightArray[j];
            j++;
        }

        k++;
    }

    while (i < leftSize)
    {
        array[k] = leftArray[i];

        i++;
        k++;
    }

    while (j < rightSize)
    {
        array[k] = rightArray[j];

        j++;
        k++;
    }
}

void mergeSort(
    int* array,
    int left,
    int right
)
{
    if (left >= right)
    {
        return;
    }

    int middle = left + (right - left) / 2;

    mergeSort(array, left, middle);
    mergeSort(array, middle + 1, right);

    merge(array, left, middle, right);
}

int binarySearch(
    const int* array,
    int left,
    int right,
    int target
)
{
    if (left > right)
    {
        return -1;
    }

    int middle = left + (right - left) / 2;

    if (array[middle] == target)
    {
        return middle;
    }

    if (target < array[middle])
    {
        return binarySearch(
            array,
            left,
            middle - 1,
            target
        );
    }

    return binarySearch(
        array,
        middle + 1,
        right,
        target
    );
}

int maxCrossingSum(
    const int* array,
    int left,
    int middle,
    int right
)
{
    int leftSum = INT_MIN;
    int sum = 0;

    for (int i = middle; i >= left; i--)
    {
        sum += array[i];

        if (sum > leftSum)
        {
            leftSum = sum;
        }
    }

    int rightSum = INT_MIN;
    sum = 0;

    for (int i = middle + 1; i <= right; i++)
    {
        sum += array[i];

        if (sum > rightSum)
        {
            rightSum = sum;
        }
    }

    return leftSum + rightSum;
}

int maxSubarraySum(
    const int* array,
    int left,
    int right
)
{
    if (left == right)
    {
        return array[left];
    }

    int middle = left + (right - left) / 2;

    int leftSum = maxSubarraySum(
        array,
        left,
        middle
    );

    int rightSum = maxSubarraySum(
        array,
        middle + 1,
        right
    );

    int crossingSum = maxCrossingSum(
        array,
        left,
        middle,
        right
    );

    return std::max(
        leftSum,
        std::max(rightSum, crossingSum)
    );
}

long long fastPower(
    long long base,
    unsigned int exponent
)
{
    if (exponent == 0)
    {
        return 1;
    }

    long long half = fastPower(
        base,
        exponent / 2
    );

    long long result = half * half;

    if (exponent % 2 != 0)
    {
        result *= base;
    }

    return result;
}

int findMax(
    const int* array,
    int left,
    int right
)
{
    if (left == right)
    {
        return array[left];
    }

    int middle = left + (right - left) / 2;

    int leftMax = findMax(
        array,
        left,
        middle
    );

    int rightMax = findMax(
        array,
        middle + 1,
        right
    );

    return std::max(leftMax, rightMax);
}

void printArray(
    const int* array,
    int size
)
{
    for (int i = 0; i < size; i++)
    {
        std::cout << array[i] << " ";
    }

    std::cout << "\n";
}

int main()
{
    int array[] = {
        8, 3, 5, 4,
        7, 6, 1, 2
    };

    int size = sizeof(array) / sizeof(array[0]);

    std::cout << "Original array: ";
    printArray(array, size);

    mergeSort(array, 0, size - 1);

    std::cout << "Sorted array:   ";
    printArray(array, size);

    int target = 6;

    int index = binarySearch(
        array,
        0,
        size - 1,
        target
    );

    std::cout
        << "Binary Search ("
        << target
        << "): index "
        << index
        << "\n";

    int subarray[] = {
        -2, 1, -3, 4,
        -1, 2, 1, -5, 4
    };

    int subarraySize =
        sizeof(subarray) / sizeof(subarray[0]);

    int maximumSum = maxSubarraySum(
        subarray,
        0,
        subarraySize - 1
    );

    std::cout
        << "Maximum subarray sum: "
        << maximumSum
        << "\n";

    std::cout
        << "Fast Power (2^10): "
        << fastPower(2, 10)
        << "\n";

    int values[] = {
        7, 2, 9, 4, 1, 8
    };

    int valuesSize =
        sizeof(values) / sizeof(values[0]);

    std::cout
        << "Maximum element: "
        << findMax(
            values,
            0,
            valuesSize - 1
        )
        << "\n";

    return 0;
}

File: algorithmic-techniques/Divide and Conquer/Python/code.py
def merge(array, left, middle, right):
    left_array = array[left:middle + 1]
    right_array = array[middle + 1:right + 1]

    i = 0
    j = 0
    k = left

    while i < len(left_array) and j < len(right_array):
        if left_array[i] <= right_array[j]:
            array[k] = left_array[i]
            i += 1
        else:
            array[k] = right_array[j]
            j += 1

        k += 1

    while i < len(left_array):
        array[k] = left_array[i]
        i += 1
        k += 1

    while j < len(right_array):
        array[k] = right_array[j]
        j += 1
        k += 1


def merge_sort(array, left, right):
    if left >= right:
        return

    middle = left + (right - left) // 2

    merge_sort(array, left, middle)
    merge_sort(array, middle + 1, right)

    merge(array, left, middle, right)


def binary_search(array, left, right, target):
    if left > right:
        return -1

    middle = left + (right - left) // 2

    if array[middle] == target:
        return middle

    if target < array[middle]:
        return binary_search(
            array,
            left,
            middle - 1,
            target
        )

    return binary_search(
        array,
        middle + 1,
        right,
        target
    )


def max_crossing_sum(array, left, middle, right):
    left_sum = float("-inf")
    total = 0

    for index in range(middle, left - 1, -1):
        total += array[index]

        if total > left_sum:
            left_sum = total

    right_sum = float("-inf")
    total = 0

    for index in range(middle + 1, right + 1):
        total += array[index]

        if total > right_sum:
            right_sum = total

    return left_sum + right_sum


def max_subarray_sum(array, left, right):
    if left == right:
        return array[left]

    middle = left + (right - left) // 2

    left_sum = max_subarray_sum(
        array,
        left,
        middle
    )

    right_sum = max_subarray_sum(
        array,
        middle + 1,
        right
    )

    crossing_sum = max_crossing_sum(
        array,
        left,
        middle,
        right
    )

    return max(
        left_sum,
        right_sum,
        crossing_sum
    )


def fast_power(base, exponent):
    if exponent == 0:
        return 1

    half = fast_power(
        base,
        exponent // 2
    )

    result = half * half

    if exponent % 2 != 0:
        result *= base

    return result


def find_max(array, left, right):
    if left == right:
        return array[left]

    middle = left + (right - left) // 2

    left_max = find_max(
        array,
        left,
        middle
    )

    right_max = find_max(
        array,
        middle + 1,
        right
    )

    return max(left_max, right_max)


def print_array(array):
    print(*array)


array = [8, 3, 5, 4, 7, 6, 1, 2]

print("Original array:", end=" ")
print_array(array)

merge_sort(array, 0, len(array) - 1)

print("Sorted array:  ", end=" ")
print_array(array)

target = 6

index = binary_search(
    array,
    0,
    len(array) - 1,
    target
)

print(f"Binary Search ({target}): index {index}")

subarray = [-2, 1, -3, 4, -1, 2, 1, -5, 4]

maximum_sum = max_subarray_sum(
    subarray,
    0,
    len(subarray) - 1
)

print("Maximum subarray sum:", maximum_sum)

print("Fast Power (2^10):", fast_power(2, 10))

values = [7, 2, 9, 4, 1, 8]

maximum = find_max(
    values,
    0,
    len(values) - 1
)

print("Maximum element:", maximum)